python/pyarrow/_s3fs.pyx

```diff
@@ -236,16 +236,24 @@ cdef class S3FileSystem(FileSystem):
             S3FileSystem(proxy_options={'scheme': 'http', 'host': 'localhost',
                                         'port': 8020, 'username': 'username',
                                         'password': 'password'})
-    allow_bucket_creation : bool, default False
+    allow_bucket_creation : boolean, default False
         Whether to allow CreateDir at the bucket-level. This option may also be
         passed in a URI query parameter.
-    allow_bucket_deletion : bool, default False
+    allow_bucket_deletion : boolean, default False
         Whether to allow DeleteDir at the bucket-level. This option may also be
         passed in a URI query parameter.
+    check_directory_existence_before_creation: boolean, default false
+        Whether to allow pessimistic directory creation in CreateDir function
+        if false, then CreateDir will try to create the directory without checking its
```

Comment:
```suggestion
        By default (False), when creating a directory, it will try to create the directory without checking its
```

I would maybe also include the "It's an optimization to try directory creation and catch the error, rather than issue two dependent I/O calls." as is explained in C++